src/Microsoft.DotNet.Build.CloudTestTasks/UploadToAzure.cs

```diff
@@ -3,94 +3,194 @@
+            }
+            catch (WebException e)
+            {
+                Log.LogError(e.Message);
+                Log.LogError(e.Status.ToString());
+                using (Stream stream = e.Response.GetResponseStream())
+                using (StreamReader errorStreamReader = new StreamReader(stream))
                 {
-                    Log.LogMessage(MessageImportance.Low, "Uploading {0} to {1}.", item.ItemSpec, blockBlob.Uri);
-                    blockBlob.UploadFromStream(fileStream);
+                    Log.LogError(errorStreamReader.ReadToEnd());
                 }
-            });
+            }
+
+            Log.LogMessage(MessageImportance.Normal, "Received response to check whether Container blobs exist");
+
+            bool result = true;
+            Parallel.ForEach(
+                Items, async (item, loopState) =>
+                    {
+                        if (ct.IsCancellationRequested)
+                        {
+                            lock (Log)
+                            {
+                                Log.LogError("Task UploadToAzure cancelled");
+                            };
+                            ct.ThrowIfCancellationRequested();
+                        }
+
+                        string relativeBlobPath = item.GetMetadata("RelativeBlobPath");
+                        if (string.IsNullOrEmpty(relativeBlobPath))
+                        {
+                            lock (Log)
+                            {
+                                Log.LogError(string.Format("Metadata 'RelativeBlobPath' is missing for item '{0}'.",item.ItemSpec));
+                            }
+                            result = false;
+                            loopState.Stop();
+                        }
+
+                        if (!File.Exists(item.ItemSpec))
+                        {
+                            lock (Log)
+                            {
+                                Log.LogError(string.Format("The file '{0}' does not exist.", item.ItemSpec));
+                            }
+                            result = false;
+                            loopState.Stop();
+                        }
+
+                        if (!Overwrite && blobsPresent.Contains(relativeBlobPath))
+                        {
+                            lock (Log)
+                            {
+                                Log.LogError(string.Format("The blob '{0}' already exists.", relativeBlobPath));
+                            }
+                            result = false;
+                            loopState.Stop();
+                        }
+
+                        lock (Log)
+                        {
+                            Log.LogMessage("Uploading {0} to {1}.", item.ItemSpec, ContainerName);
+                        }
+                        UploadClient uploadClient = new UploadClient(Log);
+                        await uploadClient.UploadBlockBlobAsync(
```

Comment:
Pass the cancellation token along, so that the outstanding web requests can be cancelled too, since those requests are things you want to cancel. Essentially, if you have a cancellation token, every Async method you call inside should also be passed the token (unless they are methods that you can't change, and don't accept tokens).

Though HttpWebRequest/Response don't have cancellable methods (HttpClient, however does), so you won't actually be able to cancel outstanding requests currently.